src/dune_rules/coq_rules.ml

```diff
@@ -564,29 +567,43 @@ let setup_rules ~sctx ~dir ~dir_contents (s : Theory.t) =
     in
     source_rule ~sctx theories
   in
-  let loc = s.buildable.loc in
-  let* () =
-    let rule =
-      let file_flags = Context.coqdoc_file_flags cctx in
-      fun mode ->
-        let* () =
-          coqdoc_rule cctx ~sctx ~mode ~theories_deps:cctx.theories_deps
-            ~name:s.name ~file_flags coq_modules
-          |> Super_context.add_rule ~loc ~dir sctx
-        in
-        Coqdoc_mode.directory mode cctx.dir name
-        |> Path.build |> Action_builder.path
-        |> Rules.Produce.Alias.add_deps (Coqdoc_mode.alias mode ~dir) ~loc
-    in
-    let* () = rule Html in
-    rule Latex
-  in
   List.concat_map coq_modules ~f:(fun m ->
       let cctx = Context.for_module cctx m in
       let { Module_rule.coqc; coqdep } = setup_rule cctx ~source_rule m in
       [ coqc; coqdep ])
   |> Super_context.add_rules ~loc ~dir sctx
 
+let setup_coqdoc_rules ~sctx ~dir ~cctx (s : Theory.t) coq_modules =
+  let loc, name = (s.buildable.loc, s.name) in
+  let rule =
+    let file_flags = Context.coqdoc_file_flags cctx in
+    fun mode ->
+      let* () =
+        coqdoc_rule cctx ~sctx ~mode ~theories_deps:cctx.theories_deps
+          ~name:s.name ~file_flags coq_modules
+        |> Super_context.add_rule ~loc ~dir sctx
+      in
+      Coqdoc_mode.directory mode cctx.dir name
+      |> Path.build |> Action_builder.path
+      |> Rules.Produce.Alias.add_deps (Coqdoc_mode.alias mode ~dir) ~loc
+  in
+  let* () = rule Html in
+  rule Latex
+
+let setup_rules ~sctx ~dir ~dir_contents (s : Theory.t) =
+  let theory =
+    let scope = SC.find_scope_by_dir sctx dir in
+    let coq_lib_db = Scope.coq_libs scope in
+    Coq_lib.DB.resolve coq_lib_db ~coq_lang_version:s.buildable.coq_lang_version
+      ~loc:s.loc s.name
+  in
+  let* cctx, coq_modules =
+    setup_cctx_and_modules ~sctx ~dir ~dir_contents s theory
+  in
+  let* () = setup_coqdoc_rules ~sctx ~dir ~cctx s coq_modules in
+  let+ () = setup_vo_rules ~sctx ~dir ~cctx s theory coq_modules in
```

Comment:
I think swapping the order here makes more sense no?

Reply:
As you wish, I did preserve the order that's in main.

Reply:
It makes sense to depend on `.glob` files after we have setup their rules IMO.

Reply:
Why?

Reply:
Doesn't really matter, but when reading the code it becomes more obvious that coqdoc doesn't need to be done before vo.

Reply:
I'm not sure you can read code generating rules to deduce what the actual rules are.

I agree aesthetically it may be more pleasing in the other way around, but AFAICT there is no real reason and in fact coqc could depend on coqdoc stuff at some point too (to embed doc as Lean does, etc)

Reply:
>in fact coqc could depend on coqdoc stuff at some point too (to embed doc as Lean does, etc)

That does not seem to be a fact at the moment. 

I'm just suggesting a readability improvement nothing more.

Reply:
I agree it makes sense as purely an aesthetical concern, however if I understood you correctly, you were claiming that "it makes sense" to generate rules in a particular order.

I was under the impression that this is not the case.

Reply:
>I'm just suggesting a readability improvement nothing more.

Reply:
I get what you mean but it is actually not an improvement in the sense that order doesn't matter at all.